src/cast_core/src/CommitLatency.cpp

```diff
@@ -0,0 +1,195 @@
+    mongocxx::options::update optionsUpdate;
+    mongocxx::options::aggregate optionsAggregate;
+    bool useSession;
+    int64_t repeat;
+    int64_t threads;
+    std::uniform_int_distribution<int64_t> amountDistribution;
+    ExecutionStrategy::RunOptions options;
+
+    PhaseConfig(PhaseContext& phaseContext, const mongocxx::database& db)
+        : collection{db[phaseContext.get<std::string>("Collection")]},
+          rp_string{phaseContext.get<std::string>("ReadPreference")},
+          useSession{phaseContext.get<bool, false>("Session")},
+          repeat{phaseContext.get<IntegerSpec>("Repeat")},
+          threads{phaseContext.get<IntegerSpec>("Threads")},
+          amountDistribution{-100, 100},
+          options{ExecutionStrategy::getOptionsFrom(phaseContext, "ExecutionsStrategy")} {
+              // write_concern
+              if( phaseContext.get<bool, false>("WriteConcernMajority") ) {
+                  wc.majority(std::chrono::milliseconds{0});
+              }
+              else {
+                  wc.nodes( (std::int32_t) phaseContext.get<IntegerSpec, false>("WriteConcern").value_or(1) );
+              }
+              if( phaseContext.get<bool, false>("WriteConcernJournal") ) {
+                  wc.journal(true);
+              }
+              if( ! phaseContext.get<bool, false>("WriteConcernJournal").value_or(true) ) {
+                  // On some mongod versions it makes a difference whether journal is unset vs explicitly set to false
+                  wc.journal(false);
+              }
+              optionsUpdate.write_concern(wc);
+
+              // read_concern
+              rc.acknowledge_string(phaseContext.get<std::string>("ReadConcern"));
+              // Ugh... read_concern cannot be set for operations individually, only through client, database, or collection.
+              // And out of those, only client is valid inside transactions. (DRIVERS-619)
+              // It seems mongocxx::database doesn't provide a getter to get the client, nor is the client available in this context.
+              // Ergo, transactions cannot be supported...
+              // TODO: Should set _client. (And options::find etc should support read_concern!)
+              collection.read_concern(rc);
+
+              // read_preference
+              if ( rp_string == "PRIMARY" ) {
+                  // TODO: This actually sends primaryPreferred to server. -> File bug in Jira.
+                  // This is benign for this test, but potentially bad for real apps!
+                  // Note that k_secondary works correctly.
+                  rp.mode(mongocxx::read_preference::read_mode::k_primary);
+              }
+              else if ( rp_string == "SECONDARY" ) {
+                  rp.mode(mongocxx::read_preference::read_mode::k_secondary);
+              }
+              else {
+                  throw InvalidConfigurationException("ReadPreference must be PRIMARY or SECONDARY.");
```

Comment:
it's not well-documented or linted yet, but we've found it's better to write this as

```c++
#include <boost/throw_exception.hpp>
....
BOOST_THROW_EXCEPTION(InvalidConfigurationException("ReadPreference must be PRIMARY or SECONDARY."));
```

since then you get line-numbers in exceptions.